grouper/services/group.py

```diff
@@ -0,0 +1,58 @@
+from typing import TYPE_CHECKING
+
+from grouper.constants import (
+    DEFAULT_ADMIN_GROUP,
+    GROUP_ADMIN,
+    PERMISSION_ADMIN,
+    PERMISSION_AUDITOR,
+    USER_ADMIN,
+)
+from grouper.entities.group import GroupJoinPolicy
+from grouper.usecases.interfaces import GroupInterface
+
+if TYPE_CHECKING:
+    from grouper.repositories.interfaces import GroupRepository, PermissionGrantRepository
+    from grouper.settings import Settings
+
+
+class GroupService(GroupInterface):
+    """High-level logic to manipulate groups."""
+
+    def __init__(self, settings, group_repository, permission_grant_repository):
+        # type: (Settings, GroupRepository, PermissionGrantRepository) -> None
+        self.settings = settings
+        self.group_repository = group_repository
+        self.permission_grant_repository = permission_grant_repository
+
+    def create_group(self, name, description, join_policy):
+        # type: (str, str, GroupJoinPolicy) -> None
+        self.group_repository.create_group(name, description, join_policy)
+
+    def grant_permission_to_group(self, permission, argument, group):
```

Comment:
I think we still need a check somewhere that the argument fits our validation regex, a la https://github.com/dropbox/merou/blob/3bcdf60c8bde6a71303299e853681b1ff433867b/grouper/permissions.py#L156

There's the same question of where it should go as in the thread on the other PR though; in the model I still have in my head it's here but if the usecase is the right place then nevermind :)

Reply:
Yeah, good call, I'll add this.